allennlp/training/trainer.py

```diff
@@ -445,670 +88,24 @@ def batch_loss(self, batch: torch.Tensor, for_training: bool) -> torch.Tensor:
-        no_val_message_template = "%s |  %8.3f  |  %8s"
-        no_train_message_template = "%s |  %8s  |  %8.3f"
-        header_template = "%s |  %-10s"
-
-        metric_names = set(train_metrics.keys())
-        if val_metrics:
-            metric_names.update(val_metrics.keys())
-
-        name_length = max([len(x) for x in metric_names])
-
-        logger.info(header_template, "Training".rjust(name_length + 13), "Validation")
-        for name in metric_names:
-            train_metric = train_metrics.get(name)
-            val_metric = val_metrics.get(name)
-
-            if val_metric is not None and train_metric is not None:
-                logger.info(dual_message_template, name.ljust(name_length), train_metric, val_metric)
-            elif val_metric is not None:
-                logger.info(no_train_message_template, name.ljust(name_length), "N/A", val_metric)
-            elif train_metric is not None:
-                logger.info(no_val_message_template, name.ljust(name_length), train_metric, "N/A")
-
-    def _validation_loss(self) -> Tuple[float, int]:
-        """
-        Computes the validation loss. Returns it and the number of batches.
-        """
-        logger.info("Validating")
-
-        self.model.eval()
-
-        if self._validation_iterator is not None:
-            val_iterator = self._validation_iterator
-        else:
-            val_iterator = self.iterator
-
-        val_generator = val_iterator(self._validation_data,
-                                     num_epochs=1,
-                                     shuffle=False)
-        num_validation_batches = val_iterator.get_num_batches(self._validation_data)
-        val_generator_tqdm = Tqdm.tqdm(val_generator,
-                                       total=num_validation_batches)
-        batches_this_epoch = 0
-        val_loss = 0
-        for batch in val_generator_tqdm:
-
-            loss = self.batch_loss(batch, for_training=False)
-            if loss is not None:
-                # You shouldn't necessarily have to compute a loss for validation, so we allow for
-                # `loss` to be None.  We need to be careful, though - `batches_this_epoch` is
-                # currently only used as the divisor for the loss function, so we can safely only
-                # count those batches for which we actually have a loss.  If this variable ever
-                # gets used for something else, we might need to change things around a bit.
-                batches_this_epoch += 1
-                val_loss += loss.detach().cpu().numpy()
-
-            # Update the description with the latest metrics
-            val_metrics = self._get_metrics(val_loss, batches_this_epoch)
-            description = self._description_from_metrics(val_metrics)
-            val_generator_tqdm.set_description(description, refresh=False)
-
-        return val_loss, batches_this_epoch
+    @classmethod
+    def from_params(cls,   # type: ignore
```

Comment:
I think this breaks backward compatibility in a pretty hard way, because the `from_params` args for your abstract `Trainer` differ from the previous concrete trainer, so if anyone has:

```
from allennlp.training import Trainer

Trainer.from_params(model, params, etc)
```
This is going to break backward compatibility. E.g if anyone has written their own `train` function, which I think is probably likely.

Reply:
that's a good point